[Datastore] Vectordb bugs and design-implementation mismatches

alxtkr77 opened the pull request:
https://iguazio.atlassian.net/browse/ML-8498
https://iguazio.atlassian.net/browse/ML-8488
https://iguazio.atlassian.net/browse/ML-8465
https://iguazio.atlassian.net/browse/ML-8464
https://iguazio.atlassian.net/browse/ML-8466

Commit:
[Datastore] Vectordb bugs and design/implementation mismatches

assaf758 commented on mlrun/artifacts/document.py:
```diff
@@ -104,20 +104,20 @@ def __new__(
             def __init__(
                 self,
-                source_path,
+                local_path,
```
please change source_path to loacal_path in the docstring

assaf758 commented on mlrun/projects/project.py:
```diff
@@ -1888,38 +1888,44 @@ def get_or_create_vector_store_collection(
-        :param tag: Version tag
         :param upload: Whether to upload the artifact
         :param labels: Key-value labels
+        :param target_path: Path to the local file
```
param description seems wrong

assaf758 reviewed (commented):
looks good, two minor comments.
Also if Chroma special logic would be replaced, consider removing it from this PR for claritiy.

iguazioauto committed:
Fix Assaf's comments

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -104,20 +104,20 @@ def __new__(
         class DynamicDocumentLoader(BaseLoader):
```
I have to say I still don't get the reasoning for using an internal-class here and why this code (the `DynamicDocumentLoader::__new__` method) cannot simply be implemented in the `MLRunLoader::__init__` method. Of course, this means that the `MLRunLoader` class will inherit from `BaseLoader` directly.

alxtkr77 replied:
We cannot inherit from BaseLoader directly, because when langchain is not installed it will generate errors

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -130,15 +130,16 @@ def lazy_load(self) -> Iterator["Document"]:  # noqa: F821
+            local_path=source_path,
             loader_spec=loader_spec,
             producer=producer,
             upload=upload,
```
Repeating my comment from previous PR (regarding the `artifact_key_instance` method), let's discuss - 
Hex values are a bit distracting here, resulting in things like `22tmp22myfile44txt` where it's hard to separate the file-name from the filled chars. Why not simply replace with -? Will result in something like -tmp-myfile-txt in this case.

alxtkr77 replied:
Then we loose uniqueness - /tmp/file and /tmp~file will generate same key

theSaarco replied:
Indeed, but that's a pretty much of a corner case, don't you think? I think it's better to have better readability for the common case than sacrifice it for some cases that may never happen.

alxtkr77 replied:
Hmm, I disagree - when such a thing happens (and it will), you won't be able to find the problem. As a result, you will have an artifact with wrong content, and worse yet, you won't even know about it.

theSaarco replied:
I have to say that I feel that users managing their files this way are bound to get confused one way or the other. BTW, this will cause issues only if they choose to upload the files to the MLRun store (otherwise the content is not handled by MLRun).

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -195,12 +196,15 @@ class DocumentArtifactSpec(ArtifactSpec):
         def __init__(
             self,
             *args,
+            document_loader=None,
+            collections=None,
+            original_source=None,
```
Please add type hints where applicable.

alxtkr77 replied:
will fix

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -216,14 +220,15 @@ def __init__(
-            self.spec.original_source = kwargs["src_path"]
+        super().__init__(**kwargs)
+        self.spec.document_loader = (
+            document_loader.to_str() if document_loader else self.spec.document_loader
```
Why `to_str()`? You're using `from_dict` to de-serialize, but in general - why not maintain the object as object? You only need to serialize it when serializing the entire object.

alxtkr77 replied:
will fix

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -216,14 +220,15 @@ def __init__(
-        key=None,
-        document_loader: DocumentLoaderSpec = DocumentLoaderSpec(),
+        original_source: Optional[str] = None,
+        document_loader: Optional[DocumentLoaderSpec] = None,
```
```suggestion
        document_loader_spec: Optional[DocumentLoaderSpec] = None,
```

alxtkr77 replied:
will fix

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -301,13 +297,19 @@ def to_langchain_documents(
-        self.spec.collections.add(collection_id)
+        if collection_id not in self.spec.collections:
+            self.spec.collections[collection_id] = "1"
+            project = mlrun.get_or_create_project(mlrun.mlconf.default_project)
+            project.update_artifact(self)
```
The artifact should not take care of updating itself in the DB. It's up to the user to update if needed through the relevant producer.
Also, this code will work in some cases, but will always assume you're registering the document through a project, while the user can log the document through the MLRun execution context (i.e. the run is the producer).

alxtkr77 replied:
will fix

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -301,13 +297,19 @@ def to_langchain_documents(
-        return self.spec.collections.discard(collection_id)
+        if collection_id in self.spec.collections:
+            self.spec.collections.pop(collection_id)
+            project = mlrun.get_or_create_project(mlrun.mlconf.default_project)
+            project.update_artifact(self)
```
Same comment here.

alxtkr77 replied:
will fix

theSaarco commented on mlrun/artifacts/manager.py:
```diff
@@ -325,7 +325,7 @@ def log_artifact(
     def update_artifact(self, producer, item: Artifact):
         self.artifact_uris[item.key] = item.uri
-        self._log_to_db(item.db_key, producer.project, producer.inputs, item)
+        self._log_to_db(item.db_key, producer.name, None, item)
```
This doesn't seem right. It will only work properly if the producer is a project, otherwise it will use the run-name as the project-name which is wrong.

alxtkr77 replied:
will fix

theSaarco reviewed (changes requested):
I have some comments on the implementation, let's discuss some of the items (for example the class structure in `MLRunLoader`.

Commit:
Addressing Saar's comments

iguazioauto committed:
Using ConfigProfile

Conflicts:
	mlrun/datastore/vectorstore.py
	tests/system/datastore/test_vectorstore.py

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -301,13 +297,38 @@ def to_langchain_documents(
+        Removes the specified collection ID from the artifact's local collection mapping.
+        This method only modifies the client-side artifact object and does not persist
+        the changes to the MLRun DB. To save the changes permanently, you must call
+        project.update_artifact() after this method.
```
```suggestion
        project.update_artifact() or context.update_artifact() after this method.
```

theSaarco commented on mlrun/artifacts/document.py:
```diff
@@ -301,13 +297,38 @@ def to_langchain_documents(
+            collection_id (str): The ID of the collection to add
+        """
+        if collection_id not in self.spec.collections:
+            self.spec.collections[collection_id] = "1"
```
I don't remember if we discussed this before - currently there's no point in having `collections` as a dict, if the value is just `"1"`. I don't think we planned to have additional data added per collection, so why do we maintain it as a dict here and not a list/set?

alxtkr77 replied:
Dictionary is a basic class we use to serialize/deserialize specs. And I just want to keep it forward compatible

theSaarco replied:
So, and this is somewhat nitpicking, but I would put `None` as value, not `"1"`.

theSaarco commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -81,22 +81,31 @@ class DatastoreProfileBasic(DatastoreProfile):
+    A profile class for managing configuration data with nested public and private attributes.
+    This class extends DatastoreProfile to handle configuration settings, separating them into
+    public and private dictionaries. Both dictionaries support nested structures, and the class
+    provides functionality to merge these attributes when needed.
```
It would be great to have a small code sample here explaining how to use this class and set credentials in it.

theSaarco commented on mlrun/datastore/vectorstore.py:
```diff
@@ -13,65 +13,39 @@
-            Raises NotImplementedError if the delete operation is not supported for the collection implementation.
+        collection_name (str): Name of the vector store collection
+        _collection_impl (VectorStore): The underlying vector store implementation
+        _mlrun_context (Union[MlrunProject, MLClientCtx]): MLRun context for artifact tracking
```
Usually we don't really document the inner workings of the class - this should be the user-facing documentation of how to use this class, not how it's built.

theSaarco commented on mlrun/datastore/vectorstore.py:
```diff
@@ -112,40 +86,66 @@ def add_documents(
+        Converts artifacts to LangChain documents, adds them to the vector store, and
+        updates the MLRun context. If documents are split, the IDs are handled appropriately.
 
         Args:
```
I see you're using the Google-style docstrings here - did you verify this renders well in Sphinx as well, for our documentation site? Most of MLRun uses different docstring format, so I'm not sure if this works or not.

theSaarco commented on mlrun/datastore/vectorstore.py:
```diff
@@ -112,40 +86,66 @@ def add_documents(
         """
         all_ids = []
-        for artifact in artifacts:
+        user_ids = kwargs.pop("ids", None)
```
Should verify that the `user_ids` collection (if exists) has the same length as the number of artifacts, otherwise the code will explode later, when looking for `user_ids[index]`. It's preferred to also verify that it is of type `list` or at least that it is iterable.

theSaarco commented on mlrun/projects/project.py:
```diff
@@ -1862,70 +1863,60 @@ def log_model(
         doc_artifact = DocumentArtifact(
             key=key,
-            document_loader=document_loader,
+            original_source=target_path or local_path,
```
```suggestion
            original_source=local_path or target_path,
```
If both are provided, you definitely want to prefer the local path.

theSaarco commented on mlrun/projects/project.py:
```diff
@@ -4502,6 +4493,25 @@ def register_datastore_profile(self, profile: DatastoreProfile):
             profile, self.name
         )
 
+    def get_config_profile(self, name: str) -> dict:
```
This is not really getting the profile, but rather is a utility to retrieve all the parameters (if you just want the profile you can call `get_datastore_profile()`). 
So, looking at it now, I actually think it's better to rename this to something like `get_config_profile_attributes` - will make it more accurate.

theSaarco commented on mlrun/execution.py:
```diff
@@ -877,28 +877,35 @@ def log_model(
         """
         doc_artifact = DocumentArtifact(
             key=key,
+            original_source=target_path or local_path,
```
```suggestion
            original_source=local_path or target_path,
```

theSaarco commented on tests/system/datastore/test_vectorstore.py:
```diff
@@ -103,66 +109,112 @@ def test_vectorstore_document_artifact(self):
+            src_name="file_path",
+            kwargs={},
+        )
+        loader = DirectoryLoader(
+            temp_dir,
+            glob="**/*.*",
+            loader_cls=MLRunLoader,
+            loader_kwargs={
+                "loader_spec": artifact_loader_spec,
+                "artifact_key": artifact_key,
+                "producer": self.project,
+                "upload": False,
```
🔥 This is awesome - it can also serve as a good usage example both for QA and for the docs.

theSaarco reviewed (changes requested):
Very nice. Have some comments and some changes needed, but overall looks very good, and I really like the tests added - they can be used as reference for tests/docs.

iguazioauto committed:
Addressing Saar's comments

theSaarco reviewed (approved):
Looks good. Have some minor comments, and for now we'll agree to disagree on the way the key is generated - but let's try to solicit feedback on that from people using the code.
